Authentication/SignUp/View/SignupController.swift

```diff
@@ -65,62 +67,120 @@ private extension SignupController {
+            } else {
+                self.signupView.passwordVisibilityButton?.hidden = false
+            }
+        })
         _viewModel.passwordValidationErrors.signal.observeNext { [unowned self] errors in
-            self._delegate.signupController(self, didFailPasswordValidationWithErrors: errors)
+            if errors.isEmpty {
+                self._delegate.signupControllerDidPassPasswordValidation(self)
+            } else {
+                self._delegate.signupController(self, didFailPasswordValidationWithErrors: errors)
+            }
         }
         if let passwordValidationMessageLabel = signupView.passwordValidationMessageLabel {
             passwordValidationMessageLabel.rex_text <~ _viewModel.passwordValidationErrors.signal.map { $0.first ?? " " }
         }
-        
-        _viewModel.passwordConfirmation <~ signupView.passwordConfirmTextField.rex_textSignal
-        _viewModel.passwordConfirmationValidationErrors.signal.observeNext { [unowned self] errors in
-            self._delegate.signupController(self, didFailPasswordConfirmationValidationWithErrors: errors)
+        if let passwordVisibilityButton = signupView.passwordVisibilityButton {
+            passwordVisibilityButton.rex_pressed.value = _viewModel.togglePasswordVisibilityCocoaAction
+            _viewModel.showPassword.signal.observeNext { [unowned self] in self.signupView.showPassword = $0 }
         }
-        if let passwordConfirmValidationMessageLabel = signupView.passwordConfirmValidationMessageLabel {
-            passwordConfirmValidationMessageLabel.rex_text <~ _viewModel.passwordConfirmationValidationErrors.signal.map { $0.first ?? " " }
+        signupView.passwordTextField.delegate = self
+        bindPasswordConfirmationElements()
+    }
+    
+    func bindPasswordConfirmationElements() {
+        if let passwordConfirmationTextField = signupView.passwordConfirmTextField {
+            _viewModel.passwordConfirmation <~ passwordConfirmationTextField.rex_textSignal.on(next: { [unowned self] text in
+                if text.isEmpty {
+                    self.signupView.passwordConfirmVisibilityButton?.hidden = true
+                } else {
+                    self.signupView.passwordConfirmVisibilityButton?.hidden = false
+                }
+                })
+            _viewModel.passwordConfirmationValidationErrors.signal.observeNext { [unowned self] errors in
+                if errors.isEmpty {
+                    self._delegate.signupControllerDidPassPasswordConfirmationValidation(self)
+                } else {
+                    self._delegate.signupController(self, didFailPasswordConfirmationValidationWithErrors: errors)
+                }
+            }
+            if let passwordConfirmValidationMessageLabel = signupView.passwordConfirmValidationMessageLabel {
+                passwordConfirmValidationMessageLabel.rex_text <~ _viewModel.passwordConfirmationValidationErrors.signal.map { $0.first ?? " " }
```

Comment:
We should have in the future a discussion of whether its okay to only show the first error. (we can leave it like this right now)